### DISCUSSION TITLE
Enabling SSH server in NixOS and connecting from (Windows) localhost?

### DISCUSSION THREAD

**Discussion by chriselsner** [nix-community/NixOS-WSL#112] [Q&A] [answered]
I'd like to enable SSH server in NixOS so that I can connect from localhost through SSH. I'm not very familiar with configuring an SSH server and using NixOS-WSL is my first time using NixOS (although I've dabbled with Nix and Home Manager before on a Mac). I've tried adding the following to my `configuration.nix`

```nix
services = {
  openssh = {
    enable = true;
    ports = [ <some-port> ]; # I think this needs to be changes from default because Windows OpenSSH is already listening on 22?
    passwordAuthentication = false;
  };
};
```

but after restarting the WSL instance I don't see `sshd` running when I run `ps aux | grep ssh`.

@nzbr  In this comment https://github.com/nix-community/NixOS-WSL/issues/31#issuecomment-849709294, I see that you were able to successfully set this up. Could you share the relevant steps and config?

**Comment by nzbr** [comment, marked as answer]
The config snippet looks correct; have you executed `nixos-rebuild switch`? Restarting the WSL VM should not be necessary in most cases

What do you need to have sshd for? If you just want to use vscode, you don't need to use SSH anymore, the regular WSL remote works just as well. I'm asking because I have found that there are better WSL-native solutions for everything I had used SSH in WSL for

**Reply by chriselsner** [reply]
I want to set up an SSH server because I'm trying to connect to WSL from my MacBook on my local network. I have multiple distros installed and ideally I would be able to ssh into all of them. After lots of reading, it seemed that the best way was to use my PC as a jumphost and to run an SSH server within each WSL distro listening on a different port. Then from my PC I can ssh into the WSL distros with `ssh chris@localhost -p <port-for-distro>`.

I actually just tried to do this again and it worked properly and I was able to connect from my Macbook. I must have used the wrong `ssh` incantation the first time. I assumed something was wrong with the SSH server because `sudo systemctl status sshd` was returning `Unit sshd.service could not be found`. However, `sudo systemctl list-units | grep ssh` returned

```
system-sshd.slice                  loaded active     active    Slice /system/sshd
sshd.socket                        loaded active     listening SSH Socket
```

which doesn't include `sshd.service` but apparently is enough.

**Reply by chriselsner** [reply]
Rerunning that command now returns

```
sshd@3-::1:60022-::1:48620.service loaded active     running   SSH Daemon ([::1]:48620)
system-sshd.slice                  loaded active     active    Slice /system/sshd
sshd.socket                        loaded active     listening SSH Socket
```

so apparently I needed to connect first?

**Reply by nzbr** [reply]
Sounds like a great way to get into WSL from a remote machine - I hadn't thought of using windows as a jump host before. That only the `sshd.socket` unit is running is completely normal. It is a systemd feature. Systemd listens on the socket and starts the corresponding service the moment something tries to connect to it, so that there are less unused services running
